servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/ConnectionPoolStrategy.java

```diff
@@ -0,0 +1,158 @@
+     * <p>
+     * The current heuristics were chosen based on a set of benchmarks under various circumstances, low connection
+     * counts, larger connection counts, low connection churn, high connection churn.
+     */
+    float RANDOM_SEARCH_FACTOR = 0.75f;
+
+    /**
+     * Select a connection from the ordered list of connections.
+     * @param connections the list of connections to pick from.
+     * @param selector a predicate used to test a connection.
+     * @param context optional {@link ContextMap} that can be used to pass information.
+     * @return the selected connection, or {@code null} if no existing connection was selected.
+     */
+    @Nullable
+    C select(List<C> connections, Predicate<C> selector, @Nullable ContextMap context);
+
+    /**
+     * A connection selection strategy that prioritizes a configurable "core" pool size and then will
+     * spill over to an overflow pool.
+     * <p>
+     * This {@link ConnectionPoolStrategy} attempts to emulate the pooling behavior often seen in thread pools.
+     * Specifically it allows for the configuration of a "core pool" size which are intended to be long-lived.
+     * Members of this core pool are selected at random that so they will get an even traffic load and, because they
+     * are equally selectable, will tend not to be removed due to idleness.
+     * <p>
+     * If the core pool cannot satisfy the load traffic can spill over to extra connections which are selected in-order.
+     * This has the property of minimizing traffic to the latest elements added outside the core pool size, thus let
+     * them tend to idle out of the pool once they're no longer of use.
+     * @param <C> the concrete type of the {@link LoadBalancedConnection}.
+     */
+    final class CorePoolStrategy<C extends LoadBalancedConnection> implements ConnectionPoolStrategy<C> {
+
+        private final int corePoolSize;
+        private final boolean forceCorePool;
+
+        CorePoolStrategy(final int corePoolSize, final boolean forceCorePool) {
+            this.corePoolSize = ensureNonNegative(corePoolSize, "corePoolSize");
+            this.forceCorePool = forceCorePool;
+        }
+
+        @Nullable
+        @Override
+        public C select(List<C> connections, Predicate<C> selector, @Nullable final ContextMap context) {
+            final int connectionCount = connections.size();
+            if (forceCorePool && connectionCount < corePoolSize) {
+                // return null so the Host will create a new connection and thus populate the connection pool.
+                return null;
+            }
+            final int randomAttempts = (int) (connectionCount * RANDOM_SEARCH_FACTOR);
+            final ThreadLocalRandom rnd = ThreadLocalRandom.current();
+            final int randomSearchSpace = min(connectionCount, corePoolSize);
+            for (int i = 0; i < randomAttempts; ++i) {
+                final C connection = connections.get(rnd.nextInt(randomSearchSpace));
```

Comment:
My assumption is that users will typically set a relatively small value for the `corePoolSize`. 
In this case, consider the following: instead of computing `randomAttempts` based on `RANDOM_SEARCH_FACTOR`, define a window for random selection, then pick a random connection within that window, if it doesn't pass selector test, continue iteration linearly for that window transitioning from the end to beginning of that window until you make `corePoolSize` attempts. If no connections are available within that window, we go linearly to test connections outside the window.

With the current approach, `(int) (0.75f * 2) == 1` . Which means we probe only a single connection from our `corePoolSize` and immediately jump into creating more connections. That is not what users will expect for the "corePoolSize". We should probe all connections inside "corePoolSize".

Reply:
That's a good point, and I like your suggestion. 👍